OTAnalytics/plugin_datastore/track_store.py

```diff
@@ -360,6 +408,66 @@ def calculate_geometries_for(
             if offset not in self._geometry_datasets.keys():
                 self._geometry_datasets[offset] = self._get_geometry_dataset_for(offset)
 
+    def apply_to_first_segments(self, consumer: Callable[[Event], None]) -> None:
+        first_detections = self._dataset.groupby(level=0, group_keys=True)
+        self._dataset["next_x"] = first_detections[track.X].shift(-1)
+        self._dataset["next_y"] = first_detections[track.Y].shift(-1)
+        first_segments: DataFrame = (
+            self._dataset.groupby(level=0, group_keys=True).head(1).copy()
+        )
+
+        as_dict = first_segments.reset_index().to_dict("index")
+
+        for value in as_dict.values():
+            event = Event(
+                road_user_id=value[track.TRACK_ID],
+                road_user_type=value[track.TRACK_CLASSIFICATION],
+                hostname=extract_hostname(value[track.VIDEO_NAME]),
+                occurrence=value[track.OCCURRENCE].to_pydatetime(),
+                frame_number=value[track.FRAME],
+                section_id=None,
+                event_coordinate=ImageCoordinate(value[track.X], value[track.Y]),
+                event_type=EventType.ENTER_SCENE,
+                direction_vector=calculate_direction_vector(
+                    value[track.X],
+                    value[track.Y],
+                    value["next_x"],
+                    value["next_y"],
+                ),
+                video_name=value[track.VIDEO_NAME],
+            )
+            consumer(event)
```

Comment:
Instead of creating an `Event` here, we could also create a `TrackSegment`. The `TrackSegment` contains the following attributes:
- track_id
- track_classification
- video_name
- occurrence (maybe for start and end detection)
- frame (maybe for start end end detection)
- start_x
- start_y
- end_x
- end_y

The `apply_to_first_segments` method would than be more general. The same belongs to `apply_to_last_segments`. What do you mean, @randy-seng ?

Reply:
Yes, that would make more sense to do it this way!